lokbot/farmer.py

```diff
@@ -821,6 +827,83 @@ def academy_farmer_thread(self, to_max_level=False):
         threading.Timer(2 * 3600, self.academy_farmer_thread, [to_max_level]).start()
         return
 
+    def _troop_training_capacity(self):
+        """
+        return total troop training capacity of all barracks
+        """
+        buildings = self.kingdom_enter.get('kingdom', {}).get('buildings', [])
+        troop_training_capacity = 0
+        for building in buildings:
+            if building.get('state') != BUILDING_STATE_NORMAL:
+                continue
+
+            if building['code'] == BUILDING_CODE_MAP['barrack']:
+                troop_training_capacity += BARRACK_LEVEL_TROOP_TRAINING_RATE_MAP[int(building['level'])]
+
+        return troop_training_capacity
+
+    def _total_troops_capacity_according_to_resources(self, troop_code):
+        """
+        return maximum number of troops according to resources
+        """
+        req_resources = TRAIN_TROOP_RESOURCE_REQUIREMENT[troop_code]
+
+        amount = 0
+        for req_resource, resource in zip(req_resources, self.resources):
+            if req_resource == 0:
+                continue
+
+            if amount == 0 or resource // req_resource <= amount:
+                amount = resource // req_resource
+
+        return amount
+
+    def _random_choice_building_position(self, building_code):
+        """
+        return position of a building type by random
+        """
+        buildings = self.kingdom_enter.get('kingdom', {}).get('buildings', [])
+        return random.choice([building['code'] for building in buildings if building['code'] == building_code])
```

Comment:
I think we need a random choice between building's **position** but not **code**, or simply return raw `building` object, and make caller to decide which property to use.

Reply:
You are right, I was not very focused